### PR TITLE
Fixed a Potential Stack Overflow Error in findPurl

### PR DESCRIPTION
**Pull request by nathannaveen**
# Description of the PR

* Fixed the potential stack overflow error for the function `findPurl` in `pkg/ingestor/parser/csaf/parser_csaf.go`
* Included unit tests for `findPurl`
* Used a visted map to solve the stack overflow, and didn't use Dynamic Programing because there was no need to go over the nodes again.

# PR Checklist

- [ ] All commits have [a Developer Certificate of Origin (DCO)](https://wiki.linuxfoundation.org/dco) -- they are generated using `-s` flag to `git commit`.
- [ ] All new changes are covered by tests
- [ ] If GraphQL schema is changed, `make generate` has been run
- [ ] If `collectsub` protobuf has been changed, `make proto` has been run
- [ ] All CI checks are passing (tests and formatting)
- [ ] All dependent PRs have already been merged

### CODE REVIEW

File: pkg/ingestor/parser/csaf/parser_csaf.go [hunk 1]
```diff
@@ -146,7 +151,7 @@ func (c *csafParser) findPkgSpec(ctx context.Context, product_id string) (*gener
 		return nil, fmt.Errorf("unable to locate product reference for id %s", product_id)
 	}
 
-	purl := findPurl(ctx, c.csaf.ProductTree, *pref)
+	purl := findPurl(ctx, c.csaf.ProductTree, *pref, make(map[string]bool))
```
Comment (lumjjb) [inline pkg/ingestor/parser/csaf/parser_csaf.go:154]: can we have this be the same and just have it call the recursive function, so rename `findPurl` to `findPurlSearch` and then
```
func findPurl(ctx, c.csaf.ProductTree, *pref) {
return findPurlSearch(ctx, c.csaf.ProductTree, *pref, make(map[string]bool))
}

Comment (nathannaveen) [reply pkg/ingestor/parser/csaf/parser_csaf.go]: Good idea, done!

**Comment by nathannaveen** [comment]
@lumjjb A friendly ping!

File: pkg/ingestor/parser/csaf/parser_csaf_test.go [hunk 2]
```diff
@@ -69,3 +72,70 @@ func Test_csafParser(t *testing.T) {
...
+				},
+				product_ref: "not equal to any tree nodes",
+			},
+			stackOverflow: true,
+			want:          nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if test.stackOverflow {
```
Comment (lumjjb) [inline pkg/ingestor/parser/csaf/parser_csaf_test.go:132]: same comment as other PR, can we make it just a regular test case and have a helper function create that test case.

**Review by lumjjb — COMMENTED** [review]
minor change request, else lgtm

File: pkg/ingestor/parser/csaf/parser_csaf_test.go (hunk 2 above)
Comment (nathannaveen) [reply pkg/ingestor/parser/csaf/parser_csaf_test.go]: Fixed, did something similar to https://github.com/guacsec/guac/pull/1146#discussion_r1294811412

**Review by lumjjb — APPROVED** [review]

**Review by pxp928 — APPROVED** [review]

**Comment by pxp928** [comment]
@nathannaveen rebase will auto-merge. Thanks!

**Comment by nathannaveen** [comment]
Closing this PR in favor of https://github.com/guacsec/guac/pull/1194 because the conflicts in this PR are extremely annoying to deal with.

**Pull request closed without merging**